feat: Add support for mutual TLS on kubewarden controller

viccuad opened the pull request:
## Description

Part of https://github.com/kubewarden/helm-charts/issues/630.

For kubewarden-controller chart, add a new `.Values.mTLS` section, where one can enable mutual TLS on the controller image.

The controller expects a ConfigMap with a Certificate Authority that is valid in the Kubernetes API server.

It will use this CA to create client certificates and keys that will be used by the policy-servers, so they can communicate via mTLS with the API server.
<!-- Please provide the link to the GitHub issue you are addressing -->


<!-- Please provide the link to the documentation related to your change, if applicable -->
<!-- [Documentation](https://<insert your url>) -->

## Test

<!-- Please provides a short description about how to test your pullrequest -->
TODO

<!--
```shell
cp <to_package_directory>
go test
```
-->

## Additional Information

### Tradeoff

<!-- Please describe, if any, the tradeoffs that you found acceptable in this pull request -->

### Potential improvement

<!-- Please describe, if any, potential improvement that you are envisioning -->

kravciak commented:
Create certificates:
```bash

FQDN=mtls.kubewarden.io

# Create CA
openssl req -nodes -batch -x509 -sha256 -days 365 -newkey rsa:2048 -keyout rootCA.key -out rootCA.crt
# Create CSR
openssl req -nodes -batch -newkey rsa:2048 -keyout domain.key -out domain.csr \
    -addext "subjectAltName = DNS:$FQDN"
# Create CRT
openssl x509 -req -CA rootCA.crt -CAkey rootCA.key -in domain.csr -out domain.crt -days 365 -CAcreateserial \
    -extfile <(echo "subjectAltName=DNS:$FQDN")
# Print CRT
openssl x509 -text -noout -in domain.crt
```

Create config files:
```yaml
# mtls/admission.yaml 
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: ValidatingAdmissionWebhook
  configuration:
    apiVersion: apiserver.config.k8s.io/v1
    kind: WebhookAdmissionConfiguration
    kubeConfigFile: "/etc/mtls/kubeconfig"
- name: MutatingAdmissionWebhook
  configuration:
    apiVersion: apiserver.config.k8s.io/v1
    kind: WebhookAdmissionConfiguration
    kubeConfigFile: "/etc/mtls/kubeconfig"

# mtls/kubeconfig 
apiVersion: v1
kind: Config
users:
- name: 'rancher-webhook.cattle-system.svc'
  user:
    client-certificate: /etc/mtls/domain.crt
    client-key: /etc/mtls/domain.key
```

Provide configuration and volume to k3s:
```bash
k3d cluster create mtls-cluster \
--k3s-arg '--kube-apiserver-arg=admission-control-config-file=/etc/mtls/admission.yaml@server:*' \
--volume "/home/username/mtls:/etc/mtls@server:*"
```

kravciak commented:
> Currently, I'm getting `"error":"failed to read client CA cert: open /tmp/k8s-webhook-server/serving-certs/client-ca.crt: no such file or directory","stacktrace"` when enabling mTLS and providing the configmap. I may be doing something incorrectly on my side..

I updated howto with complete steps. I got the same result as you.
@fabriziosestito could you try https://github.com/kubewarden/helm-charts/pull/638#issuecomment-2685640578 - I think there is a bug in controller (see workaround in the comment).

kravciak commented on charts/kubewarden-controller/values.yaml:
```diff
@@ -227,6 +227,18 @@ resources:
+  enable: false
+  # name of the ConfigMap in kubewarden-controller namespace containing the
+  configMapName: ""
+  # client CA certificate. The ConfigMap must contain a data.client-ca.crt key
```
It would be cleaner to move comments above configMapName, now comments are split in half by value.
Otherwise I tested it yesterday and it worked fine.

viccuad replied:
This is weird. Locally, I have the configMapName set to a string for testing below the code block, and when stashing the comment block it took it up. Fixed now.

kravciak commented:
@fabriziosestito found an issue with the chart (missing mount) and I am fixing an issue with otel tests

viccuad commented:
Amended the helm-chart with the correct mounting point for the upcoming controller change.

viccuad commented:
The `kubewarden-controller:latest` image now contains the needed fixes for this PR.

To summarize so far what I'm doing:

<details><summary>Toggle me</summary>




Create certs:

```bash
#!/usr/bin/env bash
set -aeEuo pipefail

FQDN=mtls.kubewarden.io

# Create CA
openssl req -nodes -batch -x509 -sha256 -days 365 -newkey rsa:2048 -keyout rootCA.key -out rootCA.crt
# Create CSR
openssl req -nodes -batch -newkey rsa:2048 -keyout domain.key -out domain.csr \
  -addext "subjectAltName = DNS:$FQDN"
# Create CRT
openssl x509 -req -CA rootCA.crt -CAkey rootCA.key -in domain.csr -out domain.crt -days 365 -CAcreateserial \
  -extfile <(echo "subjectAltName=DNS:$FQDN")
# Print CRT
openssl x509 -text -noout -in domain.crt
```

Creates `rootCA.{crt,key,srl}` and `domain.{crt,ket,srl}`.

Needed files:

```
# ./admission.yaml
---
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
  - name: ValidatingAdmissionWebhook
    configuration:
      apiVersion: apiserver.config.k8s.io/v1
      kind: WebhookAdmissionConfiguration
      kubeConfigFile: "/etc/mtls/kubeconfig"
  - name: MutatingAdmissionWebhook
    configuration:
      apiVersion: apiserver.config.k8s.io/v1
      kind: WebhookAdmissionConfiguration
      kubeConfigFile: "/etc/mtls/kubeconfig"
```
```
# ./kubeconfig
---
apiVersion: v1
kind: Config
users:
  # Config for the controller webhook servive, in the shape of
  # <controller chart name>-webhook-service.<kubewarden ns>.svc.cluster.local:443
  - name: "kubewarden-controller-webhook-service.kubewarden.svc.cluster.local:443"
    user:
      client-certificate: /etc/mtls/domain.crt
      client-key: /etc/mtls/domain.key
  # Config for all policy servers, in the shape of
  # policy-server-<policy-server name>.<policy-server ns>.svc.cluster.local:8443
  - name: "*.kubewarden.svc.cluster.local:8443"
    user:
      client-certificate: /etc/mtls/domain.crt
      client-key: /etc/mtls/domain.key
```
```
# ./mtls-configmap.yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: mtls-configmap
  namespace: kubewarden
data:
  client-ca.crt: <r rootCA.crt in base64>
```

Then, do:
```
$ k3d cluster create \
  --k3s-arg '--kube-apiserver-arg=admission-control-config-file=/etc/mtls/admission.yaml@server:*' \
  --volume "$(pwd):/etc/mtls@server:*"
```

Apply the configmap:
```
$ kubectl apply -f mtls-configmap.yml
```

Upgrade the helm chart:
```
$ helm upgrade -i --wait --namespace kubewarden --create-namespace kubewarden-controller \
  ./charts/kubewarden-controller \
  --set auditScanner.cronJob.schedule="*/2 * * * *" --set auditScanner.policyReporter=true \
  --set mTLS.enable=true --set mTLS.configMapName=mtls-configmap \
  --set image.tag=latest
```

</details>

flavio commented:
@viccuad I think this is failing because the request is sent to `https://kubewarden-controller-webhook-service.kubewarden.svc:443`, but the `kubeconfig` has this directive:

```yaml
- name: "kubewarden-controller-webhook-service.kubewarden.svc.cluster.local:443"
```

If I were you, I would drop the `.cluster.local` and try again. I think you can also omit the specify the `:443` since it's the default one for https.

viccuad commented:
Thanks! I left the :443 for completeness, indeed can be out. Updated the instructions on the other message for completion, rebased changing the order of the volumes in the template.

Policy-server itself works fine, the audit-scanner is misconfigured:
```
error":"Post \"https://policy-server-default.kubewarden.svc:8443/audit/clusterwide-ingress-policy\": remote error: tls: certificate required"
```

viccuad commented:
Updated tests and one variable assignment on the shared commit, ready for review.
